src/backend/distributed/commands/multi_copy.c

```diff
@@ -2923,3 +3070,380 @@ CopyGetAttnums(TupleDesc tupDesc, Relation rel, List *attnamelist)
+									  ALLOCSET_DEFAULT_INITSIZE,
+									  ALLOCSET_DEFAULT_MAXSIZE);
+
+	/* release finalized placement list at the end of this function */
+	MemoryContext oldContext = MemoryContextSwitchTo(localContext);
+
+	finalizedPlacementList = MasterShardPlacementList(shardId);
+
+	MemoryContextSwitchTo(oldContext);
+
+	shardState->shardId = shardId;
+	shardState->placementStateList = NIL;
+
+	foreach(placementCell, finalizedPlacementList)
+	{
+		ShardPlacement *placement = (ShardPlacement *) lfirst(placementCell);
+		CopyConnectionState *connectionState = NULL;
+		CopyPlacementState *placementState = NULL;
+		
+		MultiConnection *connection =
+			CopyGetPlacementConnection(placement, connectionStateHash, stopOnFailure);
+		if (connection == NULL)
+		{
+			failedPlacementCount++;
+			continue;
+		}
+
+		connectionState = GetConnectionState(connectionStateHash, connection);
+		placementState = palloc0(sizeof(CopyPlacementState));
+		placementState->shardState = shardState;
+		placementState->data = makeStringInfo();
+		placementState->connectionState = connectionState;
+		dlist_push_head(&connectionState->bufferedPlacementList,
+						&placementState->bufferedPlacementNode);
+		shardState->placementStateList = lappend(shardState->placementStateList,
+												 placementState);
+	}
+
+	/* if all placements failed, error out */
+	if (failedPlacementCount == list_length(finalizedPlacementList))
+	{
+		ereport(ERROR, (errmsg("could not connect to any active placements")));
+	}
+
+	/*
+	 * If stopOnFailure is true, we just error out and code execution should
+	 * never reach to this point. This is the case for reference tables and
+	 * copy from worker nodes.
+	 */
+	Assert(!stopOnFailure || failedPlacementCount == 0);
+
+	MemoryContextReset(localContext);
+}
+
+
+/*
+ * CopyGetPlacementConnection assigns a connection to the given placement. If
+ * a connection has already been assigned the placement in the current transaction
+ * then it reuses the connection. Otherwise, it requests a connection for placement.
+ */
+static MultiConnection *
+CopyGetPlacementConnection(ShardPlacement *placement, HTAB *connectionStateHash,
+						   bool stopOnFailure)
+{
+	MultiConnection *connection = NULL;
+	uint32 connectionFlags = FOR_DML;
+	char *nodeUser = CurrentUserName();
+	ShardPlacementAccess *placementAccess = NULL;
+
+	/*
+	 * Determine whether the task has to be assigned to a particular connection
+	 * due to a preceding access to the placement in the same transaction.
+	 */
+	placementAccess = CreatePlacementAccess(placement, PLACEMENT_ACCESS_DML);
+	connection = GetPlacementListConnectionIfCached(connectionFlags,
+													list_make1(placementAccess),
+													NULL);
+	if (connection != NULL)
+	{
+		return connection;
+	}
+
+	/* if execution mode set to sequential, reuse worker connection if available */
+	if (MultiShardConnectionType == SEQUENTIAL_CONNECTION)
+	{
+		connection = GetWorkerConnectionIfCached(placement->nodeName,
```

Comment:
We also need to make sure `AssignPlacementListToConnection` is called.

Reply:
> We also need to make sure `AssignPlacementListToConnection` is called.

I removed `GetWorkerConnectionIfCached()`, so I guess calls to `AssignPlacementListToConnection` are now handled in placement_connection.c.